src/backend/distributed/utils/background_jobs.c

```diff
@@ -0,0 +1,1327 @@
+				char *desiredStatusName = DatumGetCString(desiredStatusNameDatum);
+
+				ereport(ERROR,
+						(errmsg("Job reached terminal state \"%s\" instead of desired "
+								"state \"%s\"", reachedStatusName, desiredStatusName)));
+			}
+
+			/* job has reached its terminal state, done waiting */
+			break;
+		}
+
+		/* sleep for a while, before rechecking the job status */
+		CHECK_FOR_INTERRUPTS();
+		const long delay_ms = 1000;
+		(void) WaitLatch(MyLatch,
+						 WL_LATCH_SET | WL_TIMEOUT | WL_EXIT_ON_PM_DEATH,
+						 delay_ms,
+						 WAIT_EVENT_PG_SLEEP);
+
+		ResetLatch(MyLatch);
+	}
+
+	MemoryContextSwitchTo(oldContext);
+	MemoryContextDelete(waitContext);
+
+	PG_RETURN_VOID();
+}
+
+
+/*
+ * StartCitusBackgroundTaskQueueMonitor spawns a new background worker connected to the
+ * current database and owner. This background worker consumes the tasks that are ready
+ * for execution.
+ */
+BackgroundWorkerHandle *
+StartCitusBackgroundTaskQueueMonitor(Oid database, Oid extensionOwner)
+{
+	BackgroundWorker worker = { 0 };
+	BackgroundWorkerHandle *handle = NULL;
+
+	/* Configure a worker. */
+	memset(&worker, 0, sizeof(worker));
+	SafeSnprintf(worker.bgw_name, BGW_MAXLEN,
+				 "Citus Background Task Queue Monitor: %u/%u",
+				 database, extensionOwner);
+	worker.bgw_flags =
+		BGWORKER_SHMEM_ACCESS | BGWORKER_BACKEND_DATABASE_CONNECTION;
+	worker.bgw_start_time = BgWorkerStart_ConsistentState;
+
+	/* don't restart, we manage restarts from maintenance daemon */
+	worker.bgw_restart_time = BGW_NEVER_RESTART;
+	strcpy_s(worker.bgw_library_name, sizeof(worker.bgw_library_name), "citus");
+	strcpy_s(worker.bgw_function_name, sizeof(worker.bgw_library_name),
+			 "CitusBackgroundTaskQueueMonitorMain");
+	worker.bgw_main_arg = ObjectIdGetDatum(MyDatabaseId);
+	memcpy_s(worker.bgw_extra, sizeof(worker.bgw_extra), &extensionOwner,
+			 sizeof(Oid));
+	worker.bgw_notify_pid = MyProcPid;
+
+	if (!RegisterDynamicBackgroundWorker(&worker, &handle))
+	{
+		return NULL;
+	}
+
+	pid_t pid;
+	WaitForBackgroundWorkerStartup(handle, &pid);
+
+	return handle;
+}
+
+
+/*
+ * CitusBackgroundTaskQueueMonitorMain is the main entry point for the background worker
+ * running the background tasks queue monitor.
+ *
+ * It's mainloop reads a runnable task from pg_dist_background_task and progressing the
+ * tasks and jobs state machines associated with the task. When no new task can be found
+ * it will exit(0) and lets the maintenance daemon poll for new tasks.
+ *
+ * The main loop is currently implemented as a synchronous loop stepping through the task
+ * and update its state before going to the next.
+ */
+void
+CitusBackgroundTaskQueueMonitorMain(Datum arg)
+{
+	Oid databaseOid = DatumGetObjectId(arg);
+
+	/* extension owner is passed via bgw_extra */
+	Oid extensionOwner = InvalidOid;
+	memcpy_s(&extensionOwner, sizeof(extensionOwner),
+			 MyBgworkerEntry->bgw_extra, sizeof(Oid));
+
+	BackgroundWorkerUnblockSignals();
+
```

Comment:
something that would be nice to inherit from the maintenance daemon (not a PR blocker): set up error context similar to MaintenanceDaemonErrorContext, which shows where log messages came from via a CONTEXT line.

Reply:
That is not too hard, already did this for the executor. will reuse some code for that and give monitor related context.